### DISCUSSION TITLE
Preloading Container Images in Airgap

### DISCUSSION THREAD

**Discussion by jahn-swob** [siderolabs/talos#9305] [Q&A] [answered]
Is anyone currently solving for this particular issue?

In this scenario we have a completely blank environment, no existing infrastructure - aka no container registries to pull from.

In an airgapped environment this makes it difficult to get Talos going without implementing some sort of stop-gap registry for Talos to just start.

The reason why this is a bit of a conundrum is that I'd rather Talos host the registry that would then be used in standing up subsequent clusters, but in order for Talos to run, I need a registry. A bit of a circular loop here aka chicken and the egg.

If there was a way to preload a tarball of images to containerd on startup that would effectively fix this issue, but I don't think this is currently something that is offered. Is this something that is completely out of scope?

Is anyone else running into this problem or has solved this particular problem?

**Comment by smira** [comment, marked as answer]
We have some plans for Talos 1.9 to implement an image cache, but that would only cover layers, not manifests most probably in the first iteration. 

So it still requires registry access, but only to fetch manifests.

Right now I would say the easiest is to have the registry like described in the [docs](https://www.talos.dev/v1.7/advanced/air-gapped/)

**Reply by jahn-swob** [reply]
Fair enough, thank you for the reply!

**Comment by dariusaurius** [comment]
Hi, 

do you plan to import docker images, which were simply saved with "docker save" to a .tar file or is that something which is simply not possible with your solution?

**Reply by smira** [reply]
Not sure what you're asking about. The image cache creation requires a registry to pull from. Docker images are not supported in Kubernetes. But you can push your docker image to any registry, including a local one, and then image cache creation can pick it up

**Reply by dariusaurius** [reply]
Nevermind the .tar files, the image cache is so much cooler than anything I saw before. 
The perfect tool to send to sites where Linux know-how is not available. Many thanks for that! Is there a way to support you monetarily?

**Comment by smira** [comment]
Update: the [Image Cache](https://www.talos.dev/v1.9/talos-guides/configuration/image-cache/) got implemented.

[YouTube video](https://www.youtube.com/watch?v=Ir2m11Cijas).

**Reply by jahn-swob** [reply]
Awesome! Thanks for providing this feature! This is going to be huge for potential setup/bandwidth savings on initial configuration of clusters.

**Comment by steverfrancis** [comment]
Certainly!

You could always buy a support contract, or just sponsor us!

https://www.siderolabs.com/pricing/
Or

https://github.com/sponsors/siderolabs

**Reply by dariusaurius** [reply]
since we are searching for a way to deliver our containerized products to customers, I will strongly support you. Many thanks
